tools/c7n_mailer/c7n_mailer/azure/deploy.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2016-2017 Capital One Services, LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+import json
+import os
+import logging
+
+try:
+    from c7n_azure.function_package import FunctionPackage
+    from c7n_azure.template_utils import TemplateUtilities
+except ImportError:
+    FunctionPackage = TemplateUtilities = None
+    pass
+
+
+def provision(config):
+    log = logging.getLogger('c7n_mailer.azure.deploy')
+
+    function_name = config.get('function_name', 'mailer')
+
+    func_config = dict(
+        name=function_name,
+        servicePlanName=config.get('function_servicePlanName', 'cloudcustodian'),
+        location=config.get('function_location'),
+        appInsightsLocation=config.get('function_appInsightsLocation'))
+
+    template_util = TemplateUtilities()
+
+    parameters = _get_parameters(template_util, func_config)
+    group_name = parameters['servicePlanName']['value']
+    webapp_name = parameters['name']['value']
+
+    # Check if already existing
+    existing_webapp = template_util.resource_exist(group_name, webapp_name)
+
+    # Deploy
+    if not existing_webapp:
+        template_util.create_resource_group(
+            group_name, {'location': parameters['location']['value']})
+
+        template_util.deploy_resource_template(
+            group_name, 'dedicated_functionapp.json', parameters).wait()
+    else:
+        log.info("Found existing App %s (%s) in group %s" %
+                 (webapp_name, existing_webapp.location, group_name))
+
+    log.info("Building function package for %s" % webapp_name)
+
+    # Build package
+    packager = FunctionPackage(
+        function_name,
+        os.path.join(os.path.dirname(__file__), 'function.py'))
+
+    packager.build(None,
+                   entry_point=os.path.join(os.path.dirname(__file__), 'handle.py'),
+                   extra_modules={'c7n_mailer', 'ruamel'})
+
+    packager.pkg.add_contents(
+        function_name + '/config.json',
+        contents=json.dumps(config))
+
+    packager.pkg.add_contents(
+        function_name + '/function.json',
+        contents=packager.get_function_config({'mode':
+                                              {'type': 'azure-periodic',
+                                               'schedule': '0 30 * * * *'}}))
+
+    packager.close()
+
+    if packager.wait_for_status(webapp_name):
```

Comment:
Related to an earlier comment, we could move `wait_for_status` inside publish, because do the same thing in both places.

Reply:
I was thinking that maybe behavior and for sure errors would be different depending on what was happening as it gets more sophisticated though.

So like when we are deploying a new app service or plan then we'd expect to wait for a while for it to become ready, but if the app/plan already exists then you shouldn't be waiting at all, you can do a status check with a short timeout and hard fail if its not responding because something is broken.